README.md
<p align="center">
  <h1 align="center">Bit Packer</h1>
  <p align="center">Pack multiple numeric values into compact representations.</p>
  <p align="center" style="align: center;">
    <a href="https://www.npmjs.com/package/@vagabondbase/bit-packer">
      <img src="https://img.shields.io/badge/TypeScript-blue" alt="TypeScript" />
    </a>
    <a href="https://www.npmjs.com/package/@vagabondbase/bit-packer">
      <img src="https://img.shields.io/npm/v/@vagabondbase/bit-packer" alt="npm latest release" />
    </a>
    <a href="https://github.com/vagabondbase/bit-packer/actions?query=branch%3Amain">
      <img src="https://github.com/vagabondbase/bit-packer/actions/workflows/main.yml/badge.svg" alt="CI status" />
    </a>
  </p>
</p>

## Installation

```sh
npm install @vagabondbase/bit-packer
```

## Usage

You can encode an array of numbers into a `Buffer` or a `string` (hexadecimal or UTF-16) with `BitPacker.encodeArray()`:

```ts
import { BitPacker } from 'bit-packer';

// Array to buffer
const buffer = BitPacker.encodeArray([1, 2, 3]);

// Array to string
const hexString = BitPacker.encodeArray([1.234, 2.3, 3], {
  fractionDigits: 2,
  returnType: 'string-utf16',
});
```

`BitPacker.encodeArray()` accepts an array of numbers and an optional options object with the following properties:

| Property         | Type                                         | Default Value | Description                                                                       |
| ---------------- | -------------------------------------------- | ------------- | --------------------------------------------------------------------------------- |
| `returnType`     | `'string-hex' \| 'string-utf16' \| 'buffer'` | `'buffer'`    | Specifies the return type for the encoded array.                                  |
| `fractionDigits` | `number`                                     | `0`           | The number of decimal digits allowed for fractional values in the array encoding. |

You can decode either a `Buffer` or a `string` (hexadecimal or utf16le) into an array of numbers with `BitPacker.decodeArray()`:

```ts
import { BitPacker } from 'bit-packer';

// Decode from buffer
const originalArray = BitPacker.decodeArray(buffer);

// Or decode from string (hex or utf-16)
const originalArray = BitPacker.decodeArray(str);
```

**Important**: always use the `returnType` option when you need to encode to string instead of calling `toString()` on the buffer. This is because some encodings (e.g. UTF-8) may produce invalid characters and the resulting string may not be decoded correctly afterwards (e.g. the Unicode replacement character `U+FFFD`, see: [Buffer.toString()](https://nodejs.org/api/buffer.html#buffers-and-character-encodings)).

## Do you need this library?

In some cases this library may be less efficient than a simple `JSON.stringify(array)`. In other cases it may provide a **95% smaller result**.

To understand why, you only need to understand the main idea behind the library:

With 64 bits you can represent the number `1`, but you can also represent the number `99999999999999999`, which has 18 more decimal digits. This means that with the same amount of memory — by packing correctly the numbers — you could represent 19 numbers of one digit, 8 numbers of 2 digits, and so on. The efficiency of the library is inversely proportional to the maximum number of significant digits of the array numbers.

Consider, however, that algorithms such as Gzip or Brotli are extremely efficient with certain types of data, and getting a shorter string does not always lead to an overall lighter result. It is important to make decisions based on the specific data you are handling and the API available in your runtime (in some runtimes using V8 isolates it may not be obvious to be able to manually decompress data).

## Efficiency

Here's a comparison between `JSON.stringify()` and `BitPacker.encodeArray()` for a list of 1,000 random numbers with a desidered precision of 0.01.

### Comparison of string length

- `JSON.stringify()` produces a string of **7,001 characters**
- BitPacker produces a string of **673 UTF-16 characters** (-90%) or 2,692 hex characters (-62%). Hex characters are as handy to store as the output of JSON.stringify because they are just numbers and letters.

### Comparison of buffer length

- `Buffer.from(JSON.stringify())` produces **7,001 bytes** (**1510 bytes** if compressed with gzip)
- BitPacker produces **1346 bytes** (-11%), further compression is not useful

## Motivation

This library was created to solve a problem specific at VagabondBase. We wanted to handle with TypeScript **huge** amounts of data about all the tourist destinations in the world, and we wanted to use an efficient method to store lists of data with a predictable and not-too-variable number of decimal places (e.g. temperatures of a certain city over time).

We thought that making the work public was the best thing to do: the underlying algorithm and mathematics is pretty simple, but there are several edge cases to consider, and it is therefore useful to develop a library with a robust test suite.

## Want to plan a trip?

This library is developed and maintained by [VagabondBase](https://vagabondbase.com/): the all-in-one platform for travel planning.


CHANGELOG.md
# @vagabondbase/bit-packer

## 0.1.0

### Minor Changes

- 776d4eb: Allow negative `fractionDigits` values

## 0.0.3

### Patch Changes

- 88ce6e3: This changeset documents non-functional updates made to the library. The codebase itself remains unchanged, and there are no modifications to the core functionality or behavior of the library.

## 0.0.2

### Patch Changes

- fe64d99: This changeset documents non-functional updates made to the library. The codebase itself remains unchanged, and there are no modifications to the core functionality or behavior of the library.


index.test.ts
import { expect, it, test } from 'vitest';
import { encodeArray, decodeArray } from './index';

it('should encode arrays of length 1', () => {
    const arrays = [[0], [1], [2]];

    for (const array of arrays) {
        expect(decodeArray(encodeArray(array))).toEqual(array);
    }
});

it('should throw an error when encoding arrays with non-finite numbers', () => {
    expect(() => encodeArray([NaN])).toThrow();
    expect(() => encodeArray([Infinity])).toThrow();
    expect(() => encodeArray([-Infinity])).toThrow();
});

it('should throw an error when encoding arrays with non-safe integers', () => {
    expect(() => encodeArray([Number.MAX_SAFE_INTEGER + 1])).toThrow();
});

it('should throw an error when encoding arrays with non-integers', () => {
    expect(() => encodeArray([1.1])).toThrow();
});

it('should throw an error when encoding arrays with negative numbers', () => {
    expect(() => encodeArray([-1])).toThrow();
});

it('should encode and decode an array of 1M numbers', () => {
    const array = Array(1_000_000).fill(12345);
    expect(decodeArray(encodeArray(array))).toEqual(array);
});

it('should encode and decode arrays of very large numbers', () => {
    const array = Array.from({ length: 100 }, (_, i) =>
        i < 50 ? Number.MAX_SAFE_INTEGER : Number.MAX_SAFE_INTEGER - i,
    );
    expect(decodeArray(encodeArray(array))).toEqual(array);
});

it('should encode and decode arrays of numbers with same amounts of digits', () => {
    const arrays = [
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 8, 7, 6, 5, 4, 3, 2, 1],
        [11, 22, 33, 44, 55, 66, 77, 89],
    ];

    for (const array of arrays) {
        expect(decodeArray(encodeArray(array))).toEqual(array);
    }
});

it('should encode and decode arrays of negative numbers when a proper "preTransform.scale" value is provided', () => {
    const array = [-10.123456789, -20.123456789, -30.56789];
    expect(
        decodeArray(encodeArray(array, { preTransform: { scale: -100 } })),
    ).toEqual([-10.12, -20.12, -30.57]);
});

it('should encode and decode arrays of negative numbers when a proper "preTransform.translate" value is provided', () => {
    const array = [-10, -20, -30];

    expect(
        decodeArray(encodeArray(array, { preTransform: { translate: 30 } })),
    ).toEqual(array);

    expect(() =>
        encodeArray(array, { preTransform: { translate: 29 } }),
    ).toThrow();
});


src/index.test.ts
import { expect, expectTypeOf, it, test } from 'vitest';
import { encodeArray, decodeArray } from './index';
import * as zlib from 'node:zlib';

const testEncodeAndDecode = (...args: Parameters<typeof encodeArray>) => {
  const [array, options] = args;

  const hexEncoded = encodeArray(array, {
    ...options,
    returnType: 'string-hex',
  });
  const utf16Encoded = encodeArray(array, {
    ...options,
    returnType: 'string-utf16',
  });
  const buffer = encodeArray(array, { ...options, returnType: 'buffer' });
  const gzipBuffer = zlib.gzipSync(buffer);
  const deflateBuffer = zlib.deflateSync(buffer);
  const brotliBuffer = zlib.brotliCompressSync(buffer);

  expect(decodeArray(hexEncoded)).toEqual(array);
  expect(decodeArray(utf16Encoded)).toEqual(array);
  expect(decodeArray(buffer)).toEqual(array);
  expect(decodeArray(zlib.gunzipSync(gzipBuffer))).toEqual(array);
  expect(decodeArray(zlib.inflateSync(deflateBuffer))).toEqual(array);
  expect(decodeArray(zlib.brotliDecompressSync(brotliBuffer))).toEqual(array);

  for (const encoding of ['hex', 'utf16le'] as const) {
    const reEncodedBuffer = Buffer.from(buffer.toString(encoding), encoding);
    expect(decodeArray(reEncodedBuffer)).toEqual(array);
  }
};

test('Typescript types', () => {
  expectTypeOf(encodeArray([1])).toEqualTypeOf<Buffer>();

  expectTypeOf(
    encodeArray([1], { returnType: 'buffer' }),
  ).toEqualTypeOf<Buffer>();

  expectTypeOf(
    encodeArray([1], { returnType: 'string-hex' }),
  ).toEqualTypeOf<string>();

  expectTypeOf(
    encodeArray([1], { returnType: 'string-utf16' }),
  ).toEqualTypeOf<string>();
});

it('should include the correct npm version in the header', () => {
  const encoded = encodeArray([1234]);
  expect(encoded.readUInt8(0)).toEqual(
    parseInt(process.env.npm_package_version || ''),
  );
});

it('should encode arrays of length 1', () => {
  const arrays = [[0], [1], [2]];

  for (const array of arrays) {
    expect(decodeArray(encodeArray(array))).toEqual(array);
  }
});

it('should throw an error when encoding arrays with non-finite numbers', () => {
  expect(() => encodeArray([NaN])).toThrow();
  expect(() => encodeArray([Infinity])).toThrow();
  expect(() => encodeArray([-Infinity])).toThrow();
});

it('should throw an error when encoding arrays with non-safe integers', () => {
  expect(() => encodeArray([Number.MAX_SAFE_INTEGER + 1])).toThrow();
});

it('should encode and decode an array of 1M numbers', () => {
  const array = Array.from({ length: 1_000_000 }, (_, i) => i - 500_000);
  testEncodeAndDecode(array);
});

it('should encode and decode arrays of very large numbers', () => {
  const array = Array(100).fill(Number.MAX_SAFE_INTEGER);
  testEncodeAndDecode(array);
});

it('should encode and decode arrays of numbers with same amounts of digits', () => {
  const arrays = [
    [1, 2, 3, 4, 5, 6, 7, 8, 9, 8, 7, 6, 5, 4, 3, 2, 1],
    [11, 22, 33, 44, 55, 66, 77, 89],
  ];

  for (const array of arrays) {
    testEncodeAndDecode(array);
  }
});

it('should encode and decode arrays of numbers with different amounts of digits', () => {
  const array = [1, 2, 34, 567, 8999];
  testEncodeAndDecode(array);
});

it('should encode and decode arrays of numbers with different amounts of digits containing zeros', () => {
  const array = [0, 1, 10, 100, 101, 110, 111];
  testEncodeAndDecode(array);
});

it('should encode and decode arrays of negative decimal numbers', () => {
  expect(
    decodeArray(
      encodeArray([-10.12, -20.123456789, -30.56789, -40.6], {
        fractionDigits: 2,
      }),
    ),
  ).toEqual([-10.12, -20.12, -30.57, -40.6]);
});

it('should encode and decode arrays of decimal numbers and integers', () => {
  expect(
    decodeArray(
      encodeArray([1, 10, -11, -20.123456789, 30.56789, -40.6], {
        fractionDigits: 2,
      }),
    ),
  ).toEqual([1, 10, -11, -20.12, 30.57, -40.6]);
});

it('should encode and decode arrays of negative integers', () => {
  const array = [0, -1, -10, -20, -21];
  testEncodeAndDecode(array);
});

it('should encode and decode arrays of negative and positive integers', () => {
  // Do not forget that -0 is a thing in JS
  const array = [0, -1, -10, -20, -21].flatMap(n => (n === 0 ? n : [n, -n]));
  testEncodeAndDecode(array);
});

it('should encode and decode arrays with a lot of zeros', () => {
  const array1 = Array.from({ length: 101 }, (_, i) => (i + 1) * 1e6); // [1e6, 2e6, 3e6, ..., 99e6]
  const array2 = Array.from({ length: 100 }, (_, i) => (i + 1) * 1e6);
  const array3 = Array.from({ length: 99 }, (_, i) => (i + 1) * 1e6);
  testEncodeAndDecode(array1);
  testEncodeAndDecode(array2);
  testEncodeAndDecode(array3);
});


src/utils.test.ts
import { expect, it, test } from 'vitest';
import { describe } from 'vitest';
import * as utils from './utils';

describe('utils:chunk', () => {
  it('should chunk standard arrays correctly', () => {
    expect(utils.chunk({ array: [1, 2, 3, 4, 5], chunkLength: 2 })).toEqual([
      [1, 2],
      [3, 4],
      [5],
    ]);

    expect(utils.chunk({ array: [1, 2, 3, 4, 5], chunkLength: 1 })).toEqual([
      [1],
      [2],
      [3],
      [4],
      [5],
    ]);
  });

  it('should return an empty array when chunking an empty array', () => {
    expect(utils.chunk({ array: [], chunkLength: 2 })).toEqual([]);
    expect(utils.chunk({ array: [], chunkLength: 3 })).toEqual([]);
  });

  it('should throw an error when chunking an array with an invalid chunk length', () => {
    expect(() =>
      utils.chunk({ array: [1, 2, 3], chunkLength: 1 / 2 }),
    ).toThrow();

    expect(() => utils.chunk({ array: [1, 2, 3], chunkLength: -1 })).toThrow();

    expect(() => utils.chunk({ array: [1, 2, 3], chunkLength: 0 })).toThrow();

    expect(() => utils.chunk({ array: [1, 2, 3], chunkLength: NaN })).toThrow();

    expect(() =>
      utils.chunk({
        array: [1, 2, 3],
        chunkLength: Number.POSITIVE_INFINITY,
      }),
    ).toThrow();

    expect(() =>
      utils.chunk({
        array: [1, 2, 3],
        chunkLength: Number.NEGATIVE_INFINITY,
      }),
    ).toThrow();
  });
});

describe('utils:numberOfDigits', () => {
  it('should return the correct number of digits', () => {
    expect(utils.numberOfDigits(0)).toBe(1);
    expect(utils.numberOfDigits(1)).toBe(1);
    expect(utils.numberOfDigits(-1)).toBe(1);
    expect(utils.numberOfDigits(10)).toBe(2);
    expect(utils.numberOfDigits(-10)).toBe(2);
    expect(utils.numberOfDigits(100)).toBe(3);
    expect(utils.numberOfDigits(-100)).toBe(3);
  });
});

describe('utils:chunkToBigInt, utils:bigIntToChunk', () => {
  it('should return the correct bigint', () => {
    expect(utils.chunkToBigInt([1, 2, 3], { numberOfDigits: 1 })).toBe(123n);

    expect(utils.chunkToBigInt([10, 11, 20, 22], { numberOfDigits: 2 })).toBe(
      1_11_02_22n,
    );

    expect(utils.chunkToBigInt([10, 20, 300], { numberOfDigits: 3 })).toBe(
      10_020_003n,
    );

    expect(utils.chunkToBigInt([1, 20, 300], { numberOfDigits: 3 })).toBe(
      100_020_003n,
    );

    expect(
      utils.bigIntToChunk(
        utils.chunkToBigInt([1, 2, 3], { numberOfDigits: 1 }),
        { numberOfDigits: 1 },
      ),
    ).toEqual([1, 2, 3]);

    expect(
      utils.bigIntToChunk(
        utils.chunkToBigInt([10, 20, 30], { numberOfDigits: 2 }),
        { numberOfDigits: 2 },
      ),
    ).toEqual([10, 20, 30]);

    expect(
      utils.bigIntToChunk(
        utils.chunkToBigInt([1, 22, 3], { numberOfDigits: 2 }),
        { numberOfDigits: 2 },
      ),
    ).toEqual([1, 22, 3]);

    expect(
      utils.bigIntToChunk(
        utils.chunkToBigInt([1, 22, 303], { numberOfDigits: 3 }),
        { numberOfDigits: 3 },
      ),
    ).toEqual([1, 22, 303]);

    expect(
      utils.bigIntToChunk(
        utils.chunkToBigInt([10, 20, 300], {
          numberOfDigits: 3,
        }),
        { numberOfDigits: 3 },
      ),
    ).toEqual([10, 20, 300]);
  });
});


src/utils.ts
export const numberOfDigits = (number: number) =>
  number === 0 ? 1 : Math.floor(Math.log10(Math.abs(number))) + 1;

export const chunk = ({
  array,
  chunkLength,
}: {
  array: number[];
  chunkLength: number;
}) => {
  if (!Number.isSafeInteger(chunkLength) || chunkLength < 1) {
    throw new Error('The chunk length must be a finite number greater than 0');
  }

  let chunkedArray = [] as Array<typeof array>;
  for (let i = 0; i < array.length; i++) {
    const chunkIndex = Math.floor(i / chunkLength);
    if (!chunkedArray[chunkIndex]) {
      chunkedArray[chunkIndex] = [];
    }
    chunkedArray[chunkIndex].push(array[i]);
  }
  return chunkedArray;
};

// Example for [10, 20, 300] and numberOfDigits 3: ["10", "20", "300"] => ["01", "02", "003"] (reversed) => ["010", "020", "003"] (added right padding) => 10_020_003n (joined converted to bigint)
export const chunkToBigInt = (
  numbers: number[],
  { numberOfDigits: inputNumberOfDigits }: { numberOfDigits: number },
): bigint => {
  const string = numbers
    .map(n => {
      const bigInt = BigInt(n);
      const reversedString = [...bigInt.toString()].reverse().join('');
      return reversedString.padEnd(inputNumberOfDigits, '0');
    })
    .join('');

  let value = 0n;
  for (let i = 0; i < string.length; i++) {
    value += BigInt(string[i]) * 10n ** BigInt(string.length - i - 1);
  }

  return value;
};

export const bigIntToChunk = (
  bigInt: bigint,
  { numberOfDigits }: { numberOfDigits: number },
): number[] => {
  let string = bigInt.toString();
  const mod = string.length % numberOfDigits;
  if (mod !== 0) {
    string =
      Array(numberOfDigits - mod)
        .fill('0')
        .join('') + string;
  }

  const numbers = [] as number[];
  // Example for 10_020_003n with numberOfDigits = 3
  for (let i = 0; i < string.length; i += numberOfDigits) {
    const subString = [...string.slice(i, i + numberOfDigits)] // [..."010"] or [..."020"] or [..."003"]
      .reverse() // ["0", "1", "0"] or ["0", "2", "0"] or ["3", "0", "0"]
      .join(''); // "010" or "020" or "300"
    numbers.push(Number(subString)); // 10 or 20 or 300
  }
  return numbers;
};


scripts/compare-with-json.ts
// Usage: pnpm scripts:compare-with-json

import { BitPacker, EncodeArrayOptions } from '../src';
import * as zlib from 'node:zlib';

const fractionDigits = 2;
const array = Array.from({ length: 1000 }, () => Math.random());

const json = JSON.stringify(array.map(n => n.toFixed(fractionDigits)));
const jsonBuffer = Buffer.from(json);
const jsonBufferGzipped = zlib.gzipSync(jsonBuffer);

const options: EncodeArrayOptions = {
  fractionDigits,
};

const utf16 = BitPacker.encodeArray(array, {
  ...options,
  returnType: 'string-utf16',
});

const hex = BitPacker.encodeArray(array, {
  ...options,
  returnType: 'string-hex',
});

const buffer = BitPacker.encodeArray(array, {
  ...options,
});

const percentChange = (from: number, to: number) => (to / from - 1) * 100;

console.log('String Output:');
console.log(`[JSON.stringify]: ${json.length} characters`);
console.log(
  `[bit-packer]: ${hex.length} hex characters (${percentChange(
    json.length,
    hex.length,
  ).toFixed(0)}%) or ${utf16.length} UTF-16 characters (${percentChange(
    json.length,
    utf16.length,
  ).toFixed(0)}%)`,
);

console.log('\nBuffer Output:');
console.log(
  `[JSON.stringify] ${jsonBufferGzipped.length} bytes gzipped (${jsonBuffer.length} bytes uncompressed)`,
);
console.log(
  `[bit-packer] ${buffer.length} bytes (${percentChange(
    jsonBufferGzipped.length,
    buffer.length,
  ).toFixed(0)}% bytes)`,
);


src/index.ts
import { bigIntToChunk, chunk, chunkToBigInt, numberOfDigits } from './utils';

const CONSTANTS = {
  headerLength: 1 + 1 + 4 + 4,
  majorVersion: 0,
} as const;

export type EncodeArrayOptions =
  | undefined
  | {
      returnType?: 'string-hex' | 'string-utf16' | 'buffer';
      fractionDigits?: number;
    };

type EncodeArrayReturnMap<T extends EncodeArrayOptions> = T extends {
  returnType: 'string-hex';
}
  ? string
  : T extends { returnType: 'string-utf16' }
  ? string
  : Buffer;

export const encodeArray = <T extends EncodeArrayOptions>(
  inputArray: number[],
  options?: T,
) => {
  const scale = Math.pow(10, options?.fractionDigits ?? 0);

  let array = [] as typeof inputArray;

  let minValue = Math.round(inputArray[0] * scale);
  for (let i = 0; i < inputArray.length; i++) {
    const number = Math.round(inputArray[i] * scale);
    array[i] = number;
    if (number < minValue) {
      minValue = number;
    }
  }

  let translate = 0;
  if (minValue < 0) {
    translate = Math.abs(minValue) + 1;
  } else if (inputArray.includes(0)) {
    translate = 1;
  }

  for (let i = 0; i < inputArray.length; i++) {
    const number = array[i] + translate;

    if (!Number.isFinite(number)) {
      throw new Error(
        'Array must only contain finite numbers (no NaN, Infinity or -Infinity)',
      );
    }

    if (!Number.isSafeInteger(number)) {
      throw new Error(
        "Array must only contain safe integers strictly positive – or numbers that can be transformed into such through scaling and translation by the library. There's probably nothing you can do about this, besides providing a different array.",
      );
    }

    array[i] = number;
  }

  let maxNumberOfDigits = numberOfDigits(array[0]);
  for (let i = 1; i < array.length; i++) {
    const digits = numberOfDigits(array[i]);
    if (digits > maxNumberOfDigits) {
      maxNumberOfDigits = digits;
    }
  }

  // A chunk is a subarray of the original array that is encoded into a single number.
  // 19 is the number of digits in 2^64 - 1 - 1
  const chunkLength = Math.floor(19 / maxNumberOfDigits);
  const chunks = chunk({ array, chunkLength });

  const buffer = Buffer.alloc(CONSTANTS.headerLength + chunks.length * 8);
  buffer.writeUint8(CONSTANTS.majorVersion); // 1 byte for package version
  buffer.writeUint8(maxNumberOfDigits, 1); // 1 byte for maxNumberOfDigits
  buffer.writeFloatLE(scale, 2); // 4 bytes for `preTransform.scale`
  buffer.writeFloatLE(translate, 6); // 4 bytes for `preTransform.translate`

  for (let chunkIndex = 0; chunkIndex < chunks.length; chunkIndex++) {
    const chunk = chunks[chunkIndex];
    buffer.writeBigUInt64LE(
      chunkToBigInt(chunk, { numberOfDigits: maxNumberOfDigits }),
      CONSTANTS.headerLength + chunkIndex * 8,
    );
  }

  if (
    !options ||
    !('returnType' in options) ||
    options.returnType === 'buffer'
  ) {
    return buffer as EncodeArrayReturnMap<T>;
  } else {
    return buffer.toString(
      options.returnType === 'string-hex' ? 'hex' : 'utf16le',
    ) as EncodeArrayReturnMap<T>;
  }
};

export const decodeArray = (encodedArray: Buffer | string) => {
  const buffer =
    typeof encodedArray === 'string'
      ? Buffer.from(
          encodedArray,
          /^[0-9A-Fa-f]*$/.test(encodedArray) ? 'hex' : 'utf16le',
        )
      : encodedArray;

  const array: number[] = [];

  const maxNumberOfDigits = buffer.readUInt8(1);
  const scale = buffer.readFloatLE(2);
  const translate = buffer.readFloatLE(6);

  for (
    let chunkIndex = 0;
    chunkIndex < (buffer.length - CONSTANTS.headerLength) / 8;
    chunkIndex++
  ) {
    const bigIntValue = buffer.readBigUInt64LE(
      CONSTANTS.headerLength + chunkIndex * 8,
    );

    const chunk = bigIntToChunk(bigIntValue, {
      numberOfDigits: maxNumberOfDigits,
    });

    for (let i = 0; i < chunk.length; i++) {
      let value = (chunk[i] - translate) / scale;
      array.push(value);
    }
  }

  return array;
};

export const BitPacker = {
  encodeArray,
  decodeArray,
};
